{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92b02541-6242-48c8-9863-5709cdb6467d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First program "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f13c05da-b00d-449f-bc33-a0af1837adfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello,Computational Biology\n"
     ]
    }
   ],
   "source": [
    "print(\"Hello,Computational Biology\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ecf142bd-dabb-4430-8c69-1c5bdc2a1eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Variables (Which stores Data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2e694a85-89ee-4a34-9efe-ec68ec10d7ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = 10\n",
    "y = 5\n",
    "x+y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46a79b4f-aad1-4f0d-b086-aa0576c89a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n"
     ]
    }
   ],
   "source": [
    "a=20\n",
    "b = 30\n",
    "print(a+b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "be30b090-41cb-4204-8dda-bc8cb32e213f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "141\n"
     ]
    }
   ],
   "source": [
    "a=65\n",
    "b=76 \n",
    "c=a+b\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "df600271-db2c-4742-a93e-b81446a7194f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Printing and getting length of Dna "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "02d37d39-2712-4f3c-a613-afd348a8144a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence in Sample is atgcggtcccgta\n",
      "The length of Dna is  13\n"
     ]
    }
   ],
   "source": [
    "Dna =\"atgcggtcccgta\"\n",
    "print(\"The sequence in Sample is\",Dna)\n",
    "print(\"The length of Dna is \",len(Dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d3aa9195-bae5-49dd-bc3e-63c4bf235e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Strings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e975608e-e979-4788-8355-e48c7021006f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first amino acid is G\n",
      "The last codon in sequence is TCC\n",
      "The first three amino acids is GAA\n",
      "the last five amino acids is AATCC\n",
      "The middle part of sequence is  AT\n",
      "The middle part of sequence is  AT\n"
     ]
    }
   ],
   "source": [
    "Dna = \"GAATCC\"\n",
    "print(\"The first amino acid is\",Dna[0])\n",
    "print(\"The last codon in sequence is\",Dna[-3:]) \n",
    "print(\"The first three amino acids is\",Dna[0:3])\n",
    "print(\"the last five amino acids is\",Dna[-5:])\n",
    "print(\"The middle part of sequence is \",Dna[2:4])\n",
    "print(\"The middle part of sequence is \",Dna[-4:-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6cc125d-8fea-414e-8e60-a9cd358e1c2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence found in Ecoli GATCC\n",
      "Length of DNA is: 5\n",
      "The last nuceliotide is : C\n",
      "The middle part is : TC\n",
      "from last: TC\n"
     ]
    }
   ],
   "source": [
    "DNA =\"GATCC\"\n",
    "print(\"sequence found in Ecoli\",DNA)\n",
    "print(\"Length of DNA is:\",len(DNA))\n",
    "print(\"The last nuceliotide is :\",DNA[-1])\n",
    "print(\"The middle part is :\",DNA[2:4])\n",
    "print(\"from last:\",DNA[-3:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "319b63b0-f79b-4e45-bb36-06b028c8152b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PERCENTAGE OF NUCLEOTIDES CALCULATION "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67e7624d-5fd4-47c4-b88d-4596179cd987",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of sequence is: 17\n",
      "The G count in sequence is : 5\n",
      "The C count in sequence is : 3\n",
      "The GC total count in sequence is : 8\n",
      "The GC percentage is : 47.05882352941176\n"
     ]
    }
   ],
   "source": [
    "dna=\"GAATCCGTAGTAGATGC\"\n",
    "g_count=dna.count(\"G\")\n",
    "c_count=dna.count(\"C\")\n",
    "gc_total=g_count+c_count\n",
    "gc_percentage =(gc_total/len(dna))*100\n",
    "print(\"The length of sequence is:\",len(dna))\n",
    "print(\"The G count in sequence is :\",g_count)\n",
    "print(\"The C count in sequence is :\",c_count)\n",
    "print(\"The GC total count in sequence is :\",gc_total)\n",
    "print(\"The GC percentage is :\",gc_percentage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "49b92a5b-2c4b-4c08-b165-9e831d39e948",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 16\n",
      "GC%: 50.00%\n",
      "AT%: 50.00%\n"
     ]
    }
   ],
   "source": [
    "dna=\"gattcccctcgatagt\"\n",
    "dna=dna.upper()\n",
    "g_count=dna.count(\"G\")\n",
    "c_count=dna.count(\"C\")\n",
    "a_count=dna.count(\"A\")\n",
    "t_count=dna.count(\"T\")\n",
    "gc_total=g_count+c_count\n",
    "at_total=a_count+t_count\n",
    "print(\"Sequence length:\", len(dna))\n",
    "print(f\"GC%: {gc_percentage:.2f}%\")\n",
    "print(f\"AT%: {at_percentage:.2f}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "585b7fb9-e6fc-4d0d-8934-ab353b1edf14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC contentent is moderate \n"
     ]
    }
   ],
   "source": [
    "dan=\"gaattc\"\n",
    "dan=dna.upper()\n",
    "g_count=dna.count(\"G\")\n",
    "c_count=dna.count(\"C\")\n",
    "gc_total=g_count+c_count\n",
    "percentage_of_gc=(gc_total/len(dna))*100\n",
    "if gc_percentage>60:\n",
    "   print(\"GC percentage is more \")\n",
    "elif gc_percentage>40:\n",
    "   print(\"GC contentent is moderate \")\n",
    "else :\n",
    "   print(\"GC content is less\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e631d61f-81df-4863-8dfc-84c8dd645728",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking of bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f6586b53-ff2f-40ce-a6d0-ee6319628d4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "purine\n"
     ]
    }
   ],
   "source": [
    "base =\"A\"\n",
    "if base == \"A\" or base == \"T\":\n",
    "    print(\"purine\")\n",
    "else:\n",
    "    print(\"pyramidine\")\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0782e042-c89c-451a-b6c3-67923c604157",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of A : 6\n",
      "Number of T : 8\n",
      "Number of G : 8\n",
      "Number of C : 3\n",
      "Total length of dan is: 25\n",
      "Number of purines : 14\n",
      "Number of pyramidines: 11\n",
      "Percentage of purines is: 56.00000000000001\n",
      "Percentage of pyramidines is: 44.0\n"
     ]
    }
   ],
   "source": [
    "dna=\"ATGCTGTGACTGATGATCGATGATG\"\n",
    "purines=\"A\"and \"T\"\n",
    "pyramidines =\"G\" and \"C\"\n",
    "percentage_of_purines=((dna.count(\"A\")+dna.count(\"T\"))/len(dna))*100\n",
    "percentage_of_pyramidines=((dna.count(\"G\")+dna.count(\"C\"))/len(dna))*100\n",
    "print(\"Number of A :\",dna.count(\"A\"))\n",
    "print(\"Number of T :\",dna.count(\"T\"))\n",
    "print(\"Number of G :\",dna.count(\"G\"))\n",
    "print(\"Number of C :\",dna.count(\"C\"))\n",
    "print(\"Total length of dan is:\",len(dna))\n",
    "print(\"Number of purines :\",dna.count(\"A\")+dna.count(\"T\"))\n",
    "print(\"Number of pyramidines:\",dna.count(\"G\")+dna.count(\"C\"))\n",
    "print(\"Percentage of purines is:\",percentage_of_purines)\n",
    "print(\"Percentage of pyramidines is:\",percentage_of_pyramidines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3ad1db0a-a0c2-45a5-b42d-bd54dbc81315",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T\n",
      "A\n",
      "G\n",
      "C\n",
      "G\n",
      "T\n",
      "G\n"
     ]
    }
   ],
   "source": [
    "dna=\"TAGCGTG\"\n",
    "for base in dna:\n",
    "    print(base)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0498df85-fe67-49ec-86cd-a487e0d16ded",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G count: 4\n"
     ]
    }
   ],
   "source": [
    "dna = \"ATGCGTAGCG\"\n",
    "\n",
    "g_count = 0\n",
    "\n",
    "for base in dna:\n",
    "    if base == \"G\":\n",
    "        g_count += 1\n",
    "\n",
    "print(\"G count:\", g_count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f3b5fcde-e08c-4319-827f-b94749763eb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of G in sequence : 4\n"
     ]
    }
   ],
   "source": [
    "dna=\"atcgtgcatgagtga\"\n",
    "dna=dna.upper()\n",
    "a_count=0\n",
    "for base in dna:\n",
    "    if base==\"A\":\n",
    "        a_count +=1\n",
    "print(\"Number of G in sequence :\",a_count)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "52b7220c-636c-4ce3-a7d5-2eaebe206989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Complemnt: TACGCAGCATCAC\n"
     ]
    }
   ],
   "source": [
    "dna=\"atgcgtcgtagtg\"\n",
    "dna=dna.upper()\n",
    "complement={\n",
    "    \"A\" :\"T\",\n",
    "    \"G\" :\"C\",\n",
    "    \"T\" :\"A\",\n",
    "    \"C\" :\"G\"\n",
    "}\n",
    "new_dna=\"\"\n",
    "for base in dna :\n",
    "    new_dna +=complement[base]\n",
    "print(\"Complemnt:\",new_dna)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "386a6aa8-1089-4d44-9ed9-00a8433a46b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Complemnt: TACGCAGCATCAC\n",
      "Reverse Complement: CACTACGACGCAT\n"
     ]
    }
   ],
   "source": [
    "dna=\"atgcgtcgtagtg\"\n",
    "dna=dna.upper()\n",
    "complement={\n",
    "    \"A\" :\"T\",\n",
    "    \"G\" :\"C\",\n",
    "    \"T\" :\"A\",\n",
    "    \"C\" :\"G\"\n",
    "}\n",
    "new_dna=\"\"\n",
    "for base in dna :\n",
    "    new_dna +=complement[base]\n",
    "print(\"Complemnt:\",new_dna)\n",
    "reverse_complement = new_dna[::-1]\n",
    "\n",
    "print(\"Reverse Complement:\", reverse_complement)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "8ed4b0e9-753f-4388-81f0-6b9c0ed89123",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "+\n",
      "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "@READ2\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTATTG\n",
      "+\n",
      "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "@READ3\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "+\n",
      "IIIIIIII!!!!!!!!!IIIIIIIIIIIIIII\n"
     ]
    }
   ],
   "source": [
    "with open(r\"C:\\Users\\rakes\\Downloads\\patient.fastq\") as f:\n",
    "    lines = f.readlines()\n",
    "for line in lines:\n",
    "    print(line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ae05c5fe-fee2-43d7-9021-b2fd1ea3dbc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "+\n",
      "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "@READ2\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTATTG\n",
      "+\n",
      "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "@READ3\n",
      "ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "+\n",
      "IIIIIIII!!!!!!!!!IIIIIIIIIIIIIII\n"
     ]
    }
   ],
   "source": [
    "with open(r\"C:\\Users\\rakes\\Downloads\\patient.fastq\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    print(line.strip())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "27264d33-108d-48d6-93ee-083fe140be0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Header: @READ1\n",
      "Sequence: ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "Quality: IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "-----\n",
      "Header: @READ2\n",
      "Sequence: ATGACTGACCTGAGGAGAAGTCTGCCGTTATTG\n",
      "Quality: IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII\n",
      "-----\n",
      "Header: @READ3\n",
      "Sequence: ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\n",
      "Quality: IIIIIIII!!!!!!!!!IIIIIIIIIIIIIII\n",
      "-----\n"
     ]
    }
   ],
   "source": [
    "with open(r\"C:\\Users\\rakes\\Downloads\\patient.fastq\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    sequence = lines[i+1].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    print(\"Header:\", header)\n",
    "    print(\"Sequence:\", sequence)\n",
    "    print(\"Quality:\", quality)\n",
    "    print(\"-----\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f67ad627-4905-4870-bbe0-a466e68362ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1 Average Quality: 40.0\n",
      "@READ2 Average Quality: 40.0\n",
      "@READ3 Average Quality: 28.75\n"
     ]
    }
   ],
   "source": [
    "def avg_quality(q_string):\n",
    "    total = 0\n",
    "    for char in q_string:\n",
    "        total += ord(char) - 33\n",
    "    return total / len(q_string)\n",
    "\n",
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    print(header, \"Average Quality:\", round(avg_quality(quality), 2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6ba62b05-67f4-41ee-8ec2-a22f9523f79c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1 GOOD QUALITY\n",
      "@READ2 GOOD QUALITY\n",
      "@READ3 GOOD QUALITY\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    avg_q = avg_quality(quality)\n",
    "\n",
    "    if avg_q < 20:\n",
    "        print(header, \"LOW QUALITY\")\n",
    "    else:\n",
    "        print(header, \"GOOD QUALITY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "4cae2a0f-1d75-431a-b01d-a574545b3b36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1 Low Quality Bases: 0\n",
      "@READ2 Low Quality Bases: 0\n",
      "@READ3 Low Quality Bases: 9\n"
     ]
    }
   ],
   "source": [
    "def count_low_bases(q_string, threshold=20):\n",
    "    count = 0\n",
    "    for char in q_string:\n",
    "        q = ord(char) - 33\n",
    "        if q < threshold:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "\n",
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    low_bases = count_low_bases(quality)\n",
    "\n",
    "    print(header, \"Low Quality Bases:\", low_bases)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cc886da1-41b5-4480-8aa1-dbf9ce6b4716",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1 Low %: 0.0\n",
      "@READ2 Low %: 0.0\n",
      "@READ3 Low %: 28.12\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    low_bases = count_low_bases(quality)\n",
    "    percent_low = (low_bases / len(quality)) * 100\n",
    "\n",
    "    print(header, \"Low %:\", round(percent_low, 2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0b0c3876-3e0e-493d-9d0e-324bdb43fe6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@READ1 KEEP\n",
      "@READ2 KEEP\n",
      "@READ3 REMOVE\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i].strip()\n",
    "    quality = lines[i+3].strip()\n",
    "\n",
    "    low_bases = count_low_bases(quality)\n",
    "    percent_low = (low_bases / len(quality)) * 100\n",
    "\n",
    "    if percent_low > 20:\n",
    "        print(header, \"REMOVE\")\n",
    "    else:\n",
    "        print(header, \"KEEP\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a222bd3f-0248-4924-8864-a38583d28b4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cleaned FASTQ created.\n"
     ]
    }
   ],
   "source": [
    "def count_low_bases(q_string, threshold=20):\n",
    "    count = 0\n",
    "    for char in q_string:\n",
    "        q = ord(char) - 33\n",
    "        if q < threshold:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "\n",
    "cleaned_reads = []\n",
    "\n",
    "for i in range(0, len(lines), 4):\n",
    "    header = lines[i]\n",
    "    sequence = lines[i+1]\n",
    "    plus = lines[i+2]\n",
    "    quality = lines[i+3]\n",
    "\n",
    "    low_bases = count_low_bases(quality.strip())\n",
    "    percent_low = (low_bases / len(quality.strip())) * 100\n",
    "\n",
    "    if percent_low <= 20:\n",
    "        cleaned_reads.extend([header, sequence, plus, quality])\n",
    "\n",
    "with open(\"cleaned_patient.fastq\", \"w\") as f:\n",
    "    f.writelines(cleaned_reads)\n",
    "\n",
    "print(\"Cleaned FASTQ created.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ade5ebbb-c423-4fc3-8dc8-d341c09c25e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = \"ATGACTGACCTGAGGAGAAGTCTGCCGTTACTG\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "158f2cc4-9e45-4115-8692-1363b4bba685",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mutations(seq, ref):\n",
    "    mutations = []\n",
    "    for i in range(len(seq)):\n",
    "        if seq[i] != ref[i]:\n",
    "            mutations.append((i, ref[i], seq[i]))\n",
    "    return mutations\n",
    "\n",
    "\n",
    "report_data = []\n",
    "\n",
    "for i in range(0, len(cleaned_reads), 4):\n",
    "    header = cleaned_reads[i].strip()\n",
    "    sequence = cleaned_reads[i+1].strip()\n",
    "    quality = cleaned_reads[i+3].strip()\n",
    "\n",
    "    mutations = find_mutations(sequence, reference)\n",
    "\n",
    "    report_data.append({\n",
    "        \"header\": header,\n",
    "        \"mutations\": mutations\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c7bc8d79-8fc1-46c5-9088-837bf5b0c43d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final report generated.\n"
     ]
    }
   ],
   "source": [
    "report_lines = []\n",
    "\n",
    "report_lines.append(\"=== Cancer NGS Variant Report ===\\n\")\n",
    "report_lines.append(\"Patient ID: CH-2026-017\\n\\n\")\n",
    "\n",
    "for entry in report_data:\n",
    "    report_lines.append(f\"{entry['header']}\\n\")\n",
    "\n",
    "    if entry[\"mutations\"]:\n",
    "        for m in entry[\"mutations\"]:\n",
    "            report_lines.append(\n",
    "                f\"Mutation at position {m[0]}: {m[1]} -> {m[2]}\\n\"\n",
    "            )\n",
    "    else:\n",
    "        report_lines.append(\"No mutation detected.\\n\")\n",
    "\n",
    "    report_lines.append(\"\\n\")\n",
    "\n",
    "with open(\"final_report.txt\", \"w\") as f:\n",
    "    f.writelines(report_lines)\n",
    "\n",
    "print(\"Final report generated.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "ad6ead04-e486-48bc-b561-8b496ab4b248",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\rakes'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getcwd()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9f210729-90a0-45fc-9c86-d4cd61111b0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.anaconda',\n",
       " '.conda',\n",
       " '.condarc',\n",
       " '.continuum',\n",
       " '.idlerc',\n",
       " '.ipynb_checkpoints',\n",
       " '.ipython',\n",
       " '.jupyter',\n",
       " '.keras',\n",
       " '.matplotlib',\n",
       " '.spyder-py3',\n",
       " '.streamlit',\n",
       " '.vscode',\n",
       " 'alzheimers_balanced.xlsx',\n",
       " 'alz_1000samples_cpgs.parquet',\n",
       " 'alz_1000samples_csv_chunks',\n",
       " 'alz_1000samples_csv_mlready',\n",
       " 'alz_1000samples_csv_prob',\n",
       " 'alz_1000samples_imp_cpgs',\n",
       " 'alz_1000samples_parquet_chunks',\n",
       " 'alz_1000samples_with_filtered_cpgs',\n",
       " 'alz_cleaned_full.parquet',\n",
       " 'alz_merged_cleaned_prob_latest.parquet',\n",
       " 'alz_merged_cleaned_with_new_probs.csv',\n",
       " 'alz_ml_ready_reduced.csv',\n",
       " 'alz_prob_chunks.zip',\n",
       " 'anaconda3',\n",
       " 'AppData',\n",
       " 'Application Data',\n",
       " 'ARRAYS ADVANCED COADING.ipynb',\n",
       " 'capstone_model.ipynb',\n",
       " 'cleaned_patient.fastq',\n",
       " 'Contacts',\n",
       " 'Cookies',\n",
       " 'correlation_summary.csv',\n",
       " 'data',\n",
       " 'data.txt',\n",
       " 'database front end.ipynb',\n",
       " 'Documents',\n",
       " 'Downloads',\n",
       " 'Favorites',\n",
       " 'file3.txt',\n",
       " 'final_report.txt',\n",
       " 'foldability_box.png',\n",
       " 'foldability_scatter.png',\n",
       " 'Gd_vs_SwissDock_delta_vs_WT_justification_plot.png',\n",
       " 'Jedi',\n",
       " 'latest.ipynb',\n",
       " 'Links',\n",
       " 'Local Settings',\n",
       " 'ml for protein activity .ipynb',\n",
       " 'Music',\n",
       " 'My Documents',\n",
       " 'NetHood',\n",
       " 'NTUSER.DAT',\n",
       " 'ntuser.dat.LOG1',\n",
       " 'ntuser.dat.LOG2',\n",
       " 'NTUSER.DAT{2a1eba07-1a48-11f0-ade9-bdef81b03e0c}.TM.blf',\n",
       " 'NTUSER.DAT{2a1eba07-1a48-11f0-ade9-bdef81b03e0c}.TMContainer00000000000000000001.regtrans-ms',\n",
       " 'NTUSER.DAT{2a1eba07-1a48-11f0-ade9-bdef81b03e0c}.TMContainer00000000000000000002.regtrans-ms',\n",
       " 'ntuser.ini',\n",
       " 'OneDrive',\n",
       " 'outputs',\n",
       " 'packages.txt',\n",
       " 'phython codes',\n",
       " 'PrintHood',\n",
       " 'Recent',\n",
       " 'rf_top20_model.pkl',\n",
       " 'Saved Games',\n",
       " 'Searches',\n",
       " 'SendTo',\n",
       " 'stability_box.png',\n",
       " 'stability_scatter.png',\n",
       " 'Start Menu',\n",
       " 'Templates',\n",
       " 'temp_chunk_10.parquet',\n",
       " 'temp_chunk_11.parquet',\n",
       " 'temp_chunk_12.parquet',\n",
       " 'temp_chunk_13.parquet',\n",
       " 'temp_chunk_14.parquet',\n",
       " 'temp_chunk_15.parquet',\n",
       " 'temp_chunk_16.parquet',\n",
       " 'temp_chunk_17.parquet',\n",
       " 'temp_chunk_18.parquet',\n",
       " 'temp_chunk_19.parquet',\n",
       " 'temp_chunk_2.parquet',\n",
       " 'temp_chunk_20.parquet',\n",
       " 'temp_chunk_21.parquet',\n",
       " 'temp_chunk_22.parquet',\n",
       " 'temp_chunk_23.parquet',\n",
       " 'temp_chunk_24.parquet',\n",
       " 'temp_chunk_25.parquet',\n",
       " 'temp_chunk_26.parquet',\n",
       " 'temp_chunk_27.parquet',\n",
       " 'temp_chunk_28.parquet',\n",
       " 'temp_chunk_29.parquet',\n",
       " 'temp_chunk_3.parquet',\n",
       " 'temp_chunk_30.parquet',\n",
       " 'temp_chunk_31.parquet',\n",
       " 'temp_chunk_32.parquet',\n",
       " 'temp_chunk_33.parquet',\n",
       " 'temp_chunk_34.parquet',\n",
       " 'temp_chunk_35.parquet',\n",
       " 'temp_chunk_36.parquet',\n",
       " 'temp_chunk_37.parquet',\n",
       " 'temp_chunk_38.parquet',\n",
       " 'temp_chunk_39.parquet',\n",
       " 'temp_chunk_4.parquet',\n",
       " 'temp_chunk_40.parquet',\n",
       " 'temp_chunk_41.parquet',\n",
       " 'temp_chunk_42.parquet',\n",
       " 'temp_chunk_43.parquet',\n",
       " 'temp_chunk_44.parquet',\n",
       " 'temp_chunk_45.parquet',\n",
       " 'temp_chunk_46.parquet',\n",
       " 'temp_chunk_47.parquet',\n",
       " 'temp_chunk_48.parquet',\n",
       " 'temp_chunk_49.parquet',\n",
       " 'temp_chunk_5.parquet',\n",
       " 'temp_chunk_6.parquet',\n",
       " 'temp_chunk_7.parquet',\n",
       " 'temp_chunk_8.parquet',\n",
       " 'temp_chunk_9.parquet',\n",
       " 'top50_cpgs_probabilities.csv',\n",
       " 'top50_cpgs_probabilities_unique.csv',\n",
       " 'top50_cpgs_probabilities_unique_final.csv',\n",
       " 'top50_cpgs_probabilities_unique_new.csv',\n",
       " 'traindatse.ipynb',\n",
       " 'trainmap.ipynb',\n",
       " 'Untitled.ipynb',\n",
       " 'untitled.txt',\n",
       " 'Untitled1.ipynb',\n",
       " 'Untitled10.ipynb',\n",
       " 'Untitled11.ipynb',\n",
       " 'Untitled12.ipynb',\n",
       " 'Untitled13.ipynb',\n",
       " 'Untitled14.ipynb',\n",
       " 'Untitled15.ipynb',\n",
       " 'Untitled16.ipynb',\n",
       " 'Untitled17.ipynb',\n",
       " 'Untitled18.ipynb',\n",
       " 'Untitled2.ipynb',\n",
       " 'Untitled3.ipynb',\n",
       " 'Untitled4.ipynb',\n",
       " 'Untitled5.ipynb',\n",
       " 'Untitled6.ipynb',\n",
       " 'Untitled7.ipynb',\n",
       " 'Untitled8.ipynb',\n",
       " 'Untitled9.ipynb',\n",
       " 'verify_all_tables.sql',\n",
       " 'Videos',\n",
       " 'WPS Cloud Files']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79a0e30-4204-4711-9ca9-a86f1756e040",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
